file_handler.go

```diff
@@ -0,0 +1,67 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+
+	"github.com/nlopes/slack"
+)
+
+// FileHandler downloads files from slack
+type FileHandler struct {
+	SlackAPIKey          string
+	FileDownloadLocation string
+	ProxyPrefix          string
+}
+
+// Download downloads url contents to a local file
+func (handler *FileHandler) Download(file slack.File) string {
+	fileUrl := file.URLPrivate
+	if handler.FileDownloadLocation == "" {
+		return fileUrl
+	}
+	if handler.SlackAPIKey == "" {
+		return fileUrl
+	}
+	if file.IsExternal {
+		return fileUrl
+	}
+	localFileName := fmt.Sprintf("%s_%s.%s", file.ID, file.Title, file.Filetype)
+	localFilePath := filepath.Join(handler.FileDownloadLocation, localFileName)
+	go func() {
+		out, err := os.Create(localFilePath)
+		if err != nil {
+			log.Printf("Could not create file for download %s", localFilePath)
+			return
+		}
+
+		defer out.Close()
+		request, _ := http.NewRequest("GET", fileUrl, nil)
+		request.Header.Add("Authorization", "Bearer "+handler.SlackAPIKey)
+		var client = &http.Client{}
+		resp, err := client.Do(request)
+		if err != nil {
+			log.Printf("Error downloading %v", file)
+			return
+		}
+		if resp.StatusCode != http.StatusOK {
+			log.Printf("Got %d while downloading %s", resp.StatusCode, fileUrl)
+			return
+		}
+		defer resp.Body.Close()
+		_, err = io.Copy(out, resp.Body)
+		if err != nil {
+			log.Printf("Error writing %s", fileUrl)
```

Comment:
shall we also send an IRC notification in the status if something fails?

Reply:
Seems redundant if we already have it in logs, no?
If it's still desireable it would be good to also notify about other, similar errors that could happen while calling slack api throughout the codebase.

Reply:
yeah, not for this PR, but I agree that it's desirable for other events too